### PR TITLE
Adding unit tests for ApiClient

### PR DESCRIPTION
**Pull request by maciejtreder**
## Purpose
Introducing unit tests for API Client

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: tests/MigrationTool.Tests/MigrationTool.Tests.csproj [hunk 1]
```diff
@@ -11,9 +11,11 @@
     <PackageReference Include="FluentAssertions" Version="6.8.0" />
     <PackageReference Include="FluentValidation" Version="7.4.0" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.1.0" />
-    <PackageReference Include="Moq" Version="4.17.2" />
+    <PackageReference Include="Moq" Version="4.18.4" />
+    <PackageReference Include="Moq.Contrib.HttpClient" Version="1.4.0" />
     <PackageReference Include="MSTest.TestAdapter" Version="2.2.10" />
     <PackageReference Include="MSTest.TestFramework" Version="2.2.10" />
+    <PackageReference Include="NUnit" Version="3.13.3" />
```
Comment (Mielek) [inline tests/MigrationTool.Tests/MigrationTool.Tests.csproj:18]: Why do we need NUnit when we use MSTest?

File: tests/MigrationTool.Tests/Clients/ApiClientTests.cs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+        var entity = new Entity("1", EntityType.Api);
+        var policy = "some-policy";
+        var workspace = "workspace-id";
+
+        //act
+        await this.client.UploadApiPolicy(entity, policy, workspace);
+
+        //verify
+        Assert.AreEqual(request.Method, HttpMethod.Put);
+        Assert.AreEqual(request.RequestUri, $"https://management.azure.com/subscriptions/{this.subscription}/resourceGroups/{this.extractorParameters.ResourceGroup}/providers/Microsoft.ApiManagement/service/{this.extractorParameters.SourceApimName}/workspaces/{workspace}/apis/{entity.Id}/policies/policy?api-version=2022-09-01-preview");
```
Comment (Mielek) [inline tests/MigrationTool.Tests/Clients/ApiClientTests.cs]: Please consider if extracting the common service part of URI would not help in readability in all places when you check whole uri. i.e:
this.ServiceURI()+"/workspaces/{workspace}/apis/{entity.Id}/policies/policy?api-version=2022-09-01-preview"

If you only care about end path consider using the following:
`Assert.IsTrue(request.RequestUri.AbsolutePath.EndsWith($"/workspaces/{workspace}/apis/{entity.Id}/policies/policy"))`
or
`Assert.IsTrue(request.RequestUri.LocalPath.EndsWith($"/workspaces/{workspace}/apis/{entity.Id}/policies/policy"))`

if there is a need to check parameters consider following:
`Assert.AreEqual(request.RequestUri.Query, "asd=dsa&api-version=2023-03-01-preview)`

if there is a need to check both parameters and query consider following:
`Assert.AreEqual(request.RequestUri.PathAndQuery.EndsWith("asd/dsa?asd=dsa&api-version=2023-03-01-preview)`

Please as well consider adding the FluentAssertion package, to make it easier to create above assertions and have nice error messages:
`request.RequestUri.AbsolutePath.Should().EndsWith($"/workspaces/{workspace}/apis/{entity.Id}/policies/policy")`

File: tests/MigrationTool.Tests/Clients/ApiClientTests.cs [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+        list.Add(new ApiTemplateResource() { Name = "test-api-1;rev=3", OriginalName = "test-original-name-1;rev=3", Properties = new ApiProperties { DisplayName = "test-api-dn-1" } });
+
+        this.apisClient.Setup(client => client.GetAllAsync(It.IsAny<ExtractorParameters>(), It.IsAny<bool>()).Result).Returns(list);
+
+        //act
+        var result = await this.client.FetchAllApisAndVersionSets();
+
+        //verify
+        Assert.AreEqual(result.LongCount(), 1);
+        Assert.AreEqual(result.ToList()[0].ArmTemplate, list[0]); //first revision should be saved as api entity.
```
Comment (Mielek) [inline tests/MigrationTool.Tests/Clients/ApiClientTests.cs]: Please consider changing comments to error message

File: tests/MigrationTool.Tests/Helpers/ArmTemplateComparer.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+using System.Text;
+using System.Threading.Tasks;
+
+namespace MigrationTool.Tests.Helpers;
+
+public class ArmTemplateComparer : IComparer
+{
+    public int Compare(object x, object y)
+    {
+        if (((TemplateResource) x).TestEquality((TemplateResource)y)) return 0;
```
Comment (Mielek) [inline tests/MigrationTool.Tests/Helpers/ArmTemplateComparer.cs]: Consider please adding casting test like:
`x is TemplateResource resourceX && y is TemplateResource resourceY && resourceX.TestEquality(resourceY)`

File: tests/MigrationTool.Tests/Helpers/ArmExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+using Microsoft.Azure.Management.ApiManagement.ArmTemplates.Common.Templates.Abstractions;
+using System;
+
+namespace MigrationTool.Tests.Helpers;
+
+public static class ArmExtensions
+{
+    public static bool TestEquality(this TemplateResource obj, TemplateResource another)
+    {
+        return Newtonsoft.Json.JsonConvert.SerializeObject(obj).Equals(Newtonsoft.Json.JsonConvert.SerializeObject(another));
```
Comment (Mielek) [inline tests/MigrationTool.Tests/Helpers/ArmExtensions.cs]: is it possible to do `using Newtonsoft.Json` to remove the namespace here?

**Review by Mielek — APPROVED** [review]

File: tests/MigrationTool.Tests/MigrationTool.Tests.csproj (hunk 1 above)
Comment (maciejtreder) [reply tests/MigrationTool.Tests/MigrationTool.Tests.csproj:18]: We're using `NUnit` in the `FetchAllApisAndVersionSetsDifferentApis`.

`NUnit.Framework.Assert.That(result.Select(item => (item as ApiEntity).Revisions.Count), NUnit.Framework.Is.All.EqualTo(0));`

I couldn't find any `MSTest` equivalent for that.

File: tests/MigrationTool.Tests/Clients/ApiClientTests.cs (hunk 2 above)
Comment (maciejtreder) [reply tests/MigrationTool.Tests/Clients/ApiClientTests.cs]: Changed.

**Pull request merged**